### PR TITLE
Implementation for findOneAndReplace command

### LINKED ISSUE #311: Implement findOneAndReplace command
Replaces a single document based on the specified filter and sort. Returns a document in response.

`{
   "findOneAndReplace" : {
      "filter" : {filter clause},
      "sort" : {sort clause},
      "replacement" : {document to replace},
      "options" : {"upsert" : true/false, "returnDocument" : "before/after"}
   }
}`

If upsert is `true`, insert the replacement document as new record.
if returnDocument is `before` return existing document, else return replaced document.

Validation: If the replacement document "_id" field is different from document read from the DB, throw error.

### PR DESCRIPTION
**Pull request by maheshrajamani**
**What this PR does**:
Implementation for findOneAndReplace command
**Which issue(s) this PR fixes**:
Fixes #311 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java [hunk 1]
```diff
@@ -4,29 +4,111 @@
...
 import io.stargate.sgv2.jsonapi.api.model.command.clause.update.UpdateOperation;
+import io.stargate.sgv2.jsonapi.config.constants.DocumentConstants;
+import io.stargate.sgv2.jsonapi.exception.ErrorCode;
+import io.stargate.sgv2.jsonapi.exception.JsonApiException;
+import io.stargate.sgv2.jsonapi.util.JsonNodeComparator;
 import java.util.List;
 
 /** Updates the document read from the database with the updates came as part of the request. */
-public record DocumentUpdater(List<UpdateOperation> updateOperations) {
+public record DocumentUpdater(
```
Comment (maheshrajamani) [inline src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java:14]: @tatu-at-datastax I believe extending DocumentUpdater for replace, hope this fine.

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java [hunk 2]
```diff
@@ -14,6 +14,9 @@ public enum ErrorCode {
 
   DOCUMENT_UNPARSEABLE("Unable to parse the document"),
 
+  DOCUMENT_REPLACE_DIFFERENT_DOCID(
+      "The replace document and document resolved using filter has different _id"),
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCode.java:18]: has -> have

File: src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java (hunk 1 above)
Comment (tatu-at-datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java:14]: Yes, makes sense wrt functionality. If we need more types in future may want to refactor.
I'll add one idea down below however.

File: src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java [hunk 3]
```diff
@@ -4,29 +4,111 @@
...
+    // replace will be false if upsert, should never happen
+    if (docInserted) {
+      return false;
+    }
+    // Do deep clone so we can remove _id field and check
+    ObjectNode compareDoc = docToUpdate.deepCopy();
+    JsonNode idNode = compareDoc.remove(DocumentConstants.Fields.DOC_ID);
+    // The replace document cannot specify an _id value that differs from the replaced document.
+    if (replaceDocumentId != null) {
+      if (JsonNodeComparator.ascending().compare(replaceDocumentId, idNode) != 0) {
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java:90]: Simple equality check works for kinds of `JsonNode`s that are legal as doc-ids, so no need for comparator.

But maybe better is to use existing method from `JsonUtil`; `equalsOrdered()`? That would work even if allowed more types and is designed for equality checks.

Comment (maheshrajamani) [reply src/main/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdater.java:90]: OK. Will use JsonUtil

File: src/test/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdaterTest.java [hunk 4]
```diff
@@ -299,4 +299,132 @@ public void invalidSetOnParentPathWithDollar() {
               "Update operator path conflict due to overlap: 'root' ($set) vs 'root.a' ($set)");
     }
   }
+
+  @Nested
+  class ReplaceDocuemntHappy {
```
Comment (tatu-at-datastax) [inline src/test/java/io/stargate/sgv2/jsonapi/service/updater/DocumentUpdaterTest.java:304]: nit: Docuemnt -> Document

**Review by tatu-at-datastax — APPROVED** [review]
LGTM

**Pull request merged**